### PR TITLE
Added a function to do VectorUtils which swaps rows and columns of a …

### PR DESCRIPTION
**Pull request by nilsv**
…List of DFEVectors. Also replaced the same functionality in the FFT.

### CODE REVIEW

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 1]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
+	 * @param input List of DFEVectors
+	 * @param Row to colum transposed List of DFEVEctors
+	 */
+	public static <T extends KernelObjectVectorizable<T,?>>
+	List<DFEVector<T>> rowsToColumns(List<DFEVector<T>> input) {
+		List<DFEVector<T>> result = new ArrayList<DFEVector<T>>();
+		List<List<T>> list = new ArrayList<List<T>>();
+		for (int i = 0; i < input[0].getSize(); i++) {
+			list.add(new ArrayList<T>());
+			for (int j = 0; j < input.size(); j++) {
```
Comment (simontilbury) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: Suggest a foreach loop here, e.g.:
for (DFEVector<T> v : input) {
    list[i].add(v[i]);
}

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 2]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
+				list[i].add(input[j][i]);
+			}
+		}
+
+		for (int i = 0; i < input[0].getSize(); i++) {
```
Comment (simontilbury) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: Again, a foreach might make more sense, it is Java after all...

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 3]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
+	 * first DFEVector in the list and the second element of all 3 DFEVectors will end up in the second
+	 * DFEVector in the list.
+	 *
+	 * @param input List of DFEVectors
+	 * @param Row to colum transposed List of DFEVEctors
+	 */
+	public static <T extends KernelObjectVectorizable<T,?>>
+	List<DFEVector<T>> rowsToColumns(List<DFEVector<T>> input) {
+		List<DFEVector<T>> result = new ArrayList<DFEVector<T>>();
+		List<List<T>> list = new ArrayList<List<T>>();
```
Comment (simontilbury) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: Can we think of a better name than 'list'?

Comment (nilsv) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: I am open for suggestions

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 4]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
 
+	/**
+	 * Switches the dimensions of a List of DFEVecotors.
+	 * E.g. you have a list of 3 DFEVectors of size 2 then the result will be a List
+	 * of 2 DFEVectors with size 3. The first element of all 3 DFEVectors will end up in the
+	 * first DFEVector in the list and the second element of all 3 DFEVectors will end up in the second
+	 * DFEVector in the list.
+	 *
+	 * @param input List of DFEVectors
+	 * @param Row to colum transposed List of DFEVEctors
```
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: *column

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 5]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
+
+	/**
+	 * Switches the dimensions of a DFEVector of DFEVecotors.
+	 * E.g. you have a DFEVector of 3 DFEVectors of size 2 then the result will be a DFEVector
+	 * of 2 DFEVectors with size 3. The first element of all 3 DFEVectors will end up in the
+	 * first DFEVector in the outer DFEVector and the second element of
+	 * all 3 DFEVectors will end up in the second DFEVector in the outer DFEVector.
+	 *
+	 * @param input List of DFEVectors
+	 * @param Row to colum transposed List of DFEVEctors
```
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: *column

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 6]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
+		}
+
+		for (List<T> l : list) {
+			result.add(DFEVectorType.newInstance(l));
+		}
+		return result;
+	}
+
+	/**
+	 * Switches the dimensions of a DFEVector of DFEVecotors.
```
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: typo

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 7]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
 		return DFEVectorType.newInstance(output);
 	}
 
+	/**
+	 * Switches the dimensions of a List of DFEVecotors.
```
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: and again

File: src/maxpower/kernel/utils/VectorUtils.maxj (hunk 4 above)
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: *DFEVectors

File: src/maxpower/kernel/utils/VectorUtils.maxj (hunk 5 above)
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: DFEVectors again. Also, this returns a DFEVector, not a list.

File: src/maxpower/kernel/utils/VectorUtils.maxj [hunk 8]
```diff
@@ -87,4 +87,46 @@ public class VectorUtils {
...
+		for (int i = 0; i < input[0].getSize(); i++) {
+			list.add(new ArrayList<T>());
+			for (DFEVector<T> v : input) {
+				list[i].add(v[i]);
+			}
+		}
+
+		for (List<T> l : list) {
+			result.add(DFEVectorType.newInstance(l));
+		}
```
Comment (chr1sj0nes) [inline src/maxpower/kernel/utils/VectorUtils.maxj]: You can merge these two loops and avoid creating the list of lists. i.e.:
```
for (int i = 0; i < input[0].getSize(); i++) {
  List<T> elements = new ArrayList<T>(input.size());
  for (DFEVector<T> v : input) {
    elements.add(v[i]);
  }
  result.add(DFEVectorType.newInstance(elements));
}
```

**Pull request merged**